Source/NSubstitute/Routing/Handlers/ReturnAutoValue.cs

```diff
@@ -13,37 +13,122 @@ public enum AutoValueBehaviour
+        public ReturnAutoValue(Func<AutoValueBehaviour> autoValueBehaviour, IAutoValueProvider autoValueProvider, IConfigureCall configureCall)
         {
-            _autoValueProviders = autoValueProviders;
+            _autoValueProvider = autoValueProvider;
             ConfigureCall = configureCall;
             _autoValueBehaviour = autoValueBehaviour;
         }
 
         public RouteAction Handle(ICall call)
         {
             var type = call.GetReturnType();
-            var compatibleProviders = _autoValueProviders.Where(x => x.CanProvideValueFor(type)).FirstOrNothing();
+            var compatibleProviders = _autoValueProvider.GetValue(type);
             return compatibleProviders.Fold(
-                RouteAction.Continue,
-                ReturnValueUsingProvider(call, type));
+                () => NoReturnValue(call),
+                x => ReturnValue(call, x));
         }
 
-        private Func<IAutoValueProvider, RouteAction> ReturnValueUsingProvider(ICall call, Type type)
+        private IEnumerable<ByRefArgument> GetByRefValues(CallInfo args)
         {
-            return provider =>
+            var parameters =
+                args.ArgTypes()
+                    .Select((x, i) => new { Parameter = x, Index = i })
+                    .Where(x => x.Parameter.IsByRef);
+
+            foreach (var p in parameters)
+            {
+                var type = p.Parameter;
+                var value = _autoValueProvider.GetValue(type);
+
+                if (value.HasValue())
+                    yield return new ByRefArgument(
+                        p.Index,
+                        new Lazy<object>(() => value.ValueOrDefault()));
+            }
+        }
+
+        private void SetByRefValues(CallInfo callInfo, IEnumerable<ByRefArgument> values)
+        {
+            foreach (var value in values)
+                callInfo[value.Index] = value.Value.Value;
+        }
+
+        private ReturnValueFromFunc<T> GetReturnValue<T>(T value, IEnumerable<ByRefArgument> byRefValues)
+        {
+            return new ReturnValueFromFunc<T>(callInfo =>
+            {
+                SetByRefValues(callInfo, byRefValues);
+                return value;
+            });
+        }
+
+        private RouteAction NoReturnValue(ICall call)
+        {
+            var callInfo = new CallInfoFactory().Create(call);
+            var byRefValues = 
+                GetByRefValues(callInfo)
+                .Where(x => !WasModified(x, call))
+                .ToArray();
+
+            if (byRefValues.Length == 0)
+                return RouteAction.Continue();
+
+            if (_autoValueBehaviour() == AutoValueBehaviour.UseValueForSubsequentCalls)
+                ConfigureCall.SetResultForCall(call,
+                    GetReturnValue<object>(null, byRefValues),
+                    MatchArgs.AsSpecifiedInCall);
```

Comment:
We're safe here because `NSubstitute` will always override the result and return the last value setup.

In this case specifically, `second` would be `value` and the `ref` arg would be `default(T)`.